### PR TITLE
[misc] Make stack traceback slimer for kernels and functions

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----
```py
import taichi as ti

@ti.func
def hello1():
    print(a)

@ti.func
def hello2():
    hello1()

@ti.func
def hello3():
    hello2()

@ti.kernel
def main():
    hello3()

main()
```
Before:
```
[Taichi] mode=development
[Taichi] preparing sandbox at /tmp/taichi-fstwqyo_
[Taichi] <dev mode>, llvm 10.0.0, commit e9cecb48, python 3.8.3
Traceback (most recent call last):
  File "stack.py", line 19, in <module>
    main()
  File "/root/taichi/python/taichi/lang/kernel.py", line 528, in wrapped
    return primal(*args, **kwargs)
  File "/root/taichi/python/taichi/lang/kernel.py", line 459, in __call__
    self.materialize(key=key, args=args, arg_features=arg_features)
  File "/root/taichi/python/taichi/lang/kernel.py", line 339, in materialize
    taichi_kernel = taichi_kernel.define(taichi_ast_generator)
  File "/root/taichi/python/taichi/lang/kernel.py", line 336, in taichi_ast_generator
    compiled()
  File "stack.py", line 17, in main
    hello3()
  File "/root/taichi/python/taichi/lang/kernel.py", line 37, in decorated
    return func.__call__(*args)
  File "/root/taichi/python/taichi/lang/kernel.py", line 54, in __call__
    ret = self.compiled(*args)
  File "stack.py", line 13, in hello3
    hello2()
  File "/root/taichi/python/taichi/lang/kernel.py", line 37, in decorated
    return func.__call__(*args)
  File "/root/taichi/python/taichi/lang/kernel.py", line 54, in __call__
    ret = self.compiled(*args)
  File "stack.py", line 9, in hello2
    hello1()
  File "/root/taichi/python/taichi/lang/kernel.py", line 37, in decorated
    return func.__call__(*args)
  File "/root/taichi/python/taichi/lang/kernel.py", line 54, in __call__
    ret = self.compiled(*args)
  File "stack.py", line 5, in hello1
    print(a)
NameError: name 'a' is not defined
```
After:
```
[Taichi] mode=development
[Taichi] preparing sandbox at /tmp/taichi-sg3gbs_w
[Taichi] <dev mode>, llvm 10.0.0, commit e9cecb48, python 3.8.3
Traceback (most recent call last):
  File "stack.py", line 19, in <module>
    main()
  File "/root/taichi/python/taichi/lang/kernel.py", line 457, in __call__
    self.materialize(key=key, args=args, arg_features=arg_features)
  File "/root/taichi/python/taichi/lang/kernel.py", line 337, in materialize
    taichi_kernel = taichi_kernel.define(taichi_ast_generator)
  File "/root/taichi/python/taichi/lang/kernel.py", line 334, in taichi_ast_generator
    compiled()
  File "stack.py", line 17, in main
    hello3()
  File "/root/taichi/python/taichi/lang/kernel.py", line 49, in __call__
    return self.compiled(*args)
  File "stack.py", line 13, in hello3
    hello2()
  File "/root/taichi/python/taichi/lang/kernel.py", line 49, in __call__
    return self.compiled(*args)
  File "stack.py", line 9, in hello2
    hello1()
  File "/root/taichi/python/taichi/lang/kernel.py", line 49, in __call__
    return self.compiled(*args)
  File "stack.py", line 5, in hello1
    print(a)
NameError: name 'a' is not defined
```

### CODE REVIEW

File: python/taichi/lang/kernel.py [hunk 1]
```diff
@@ -51,8 +46,11 @@ def __init__(self, func, classfunc=False):
     def __call__(self, *args):
         if self.compiled is None:
             self.do_compile()
-        ret = self.compiled(*args)
-        return ret
+        return self.compiled(*args)
+
+    def compiled(self, *args):
+        self.do_compile()
```
Comment (k-ye) [inline python/taichi/lang/kernel.py:52]: Is this infinite recursion?

Comment (archibate) [reply python/taichi/lang/kernel.py:52]: Didn't notice, sorry!

**Comment by Rullec** [comment]
KGTM!

**Comment by archibate** [comment]
Closing now, I think we need a bit more discussion on how to making stack traceback intutitive & human-readable.

**Pull request closed without merging**